### PR TITLE
Add verf ported from SIMDPirates for Julia 1.6

### PR DESCRIPTION
**Pull request by brenhinkeller**
After looking at both packages (this and SLEEFPirates) this ended up seeming most at home with the other remaining llvmcall-based v* functions here as far as I could tell, but it could go either way. Let me know if this seems like a reasonable place to put it!

### CODE REVIEW

File: src/llvm_intrin/verf.jl [hunk 1]
```diff
@@ -0,0 +1,607 @@
...
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+@inline function verf(v::Vec{8,Float64})
+    llvmcall(("""
+    attributes #0 = { alwaysinline nounwind readnone speculatable }
```
Comment (chriselrod) [inline src/llvm_intrin/verf.jl:32]: I was getting an error about callsite `speculatable`s.

File: src/llvm_intrin/verf.jl [hunk 2]
```diff
@@ -0,0 +1,607 @@
...
+      %102 = xor <8 x i64> %100, <i64 -9223372036854775808, i64 -9223372036854775808, i64 -9223372036854775808, i64 -9223372036854775808, i64 -9223372036854775808, i64 -9223372036854775808, i64 -9223372036854775808, i64 -9223372036854775808>
+      %103 = bitcast <8 x i64> %102 to <8 x double>
+      %104 = select <8 x i1> %101, <8 x double> %103, <8 x double> %99
+      %105 = select <8 x i1> %23, <8 x double> %84, <8 x double> %104
+      br label %106
+    ; 106:                                              ; preds = %9, %58, %83
+      %107 = phi <8 x double> [ %19, %9 ], [ %105, %83 ], [ %79, %58 ]
+      ret <8 x double> %107
+    }
+    """, "entry"), Vec{8,Float64}, Tuple{Vec{8,Float64}}, v)
```
Comment (chriselrod) [inline src/llvm_intrin/verf.jl:154]: You need to unwrap and re-wrap the `Vec`s.
Replacing the last line with
```julia
   """, "entry"), _Vec{8,Float64}, Tuple{_Vec{8,Float64}}, data(v)) |> Vec
```
should work.

File: src/llvm_intrin/verf.jl [hunk 3]
```diff
@@ -0,0 +1,607 @@
...
+end
+
+@inline function verf(v::Vec{4,Float64})
+    llvmcall(("""
+    attributes #7 = { alwaysinline nounwind readnone }
+    attributes #8 = { alwaysinline nounwind readnone speculatable }
```
Comment (chriselrod) [inline src/llvm_intrin/verf.jl:160]: Same thing for all the other definitions.

File: src/llvm_intrin/verf.jl [hunk 4]
```diff
@@ -0,0 +1,607 @@
...
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
```
Comment (chriselrod) [inline src/llvm_intrin/verf.jl:29]: Mind adding a `if AVX512F` guard here?

File: src/llvm_intrin/verf.jl [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+      %104 = select <8 x i1> %101, <8 x double> %103, <8 x double> %99
+      %105 = select <8 x i1> %23, <8 x double> %84, <8 x double> %104
+      br label %106
+    ; 106:                                              ; preds = %9, %58, %83
+      %107 = phi <8 x double> [ %19, %9 ], [ %105, %83 ], [ %79, %58 ]
+      ret <8 x double> %107
+    }
+    """, "entry"), Vec{8,Float64}, Tuple{Vec{8,Float64}}, v)
+end
+
```
Comment (chriselrod) [inline src/llvm_intrin/verf.jl:156]: mind adding a `if AVX` around the definition?

File: src/llvm_intrin/verf.jl [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+      %113 = select <4 x i1> %26, <4 x double> %92, <4 x double> %112
+      br label %114
+    ; 114:                                              ; preds = %11, %64, %91
+      %115 = phi <4 x double> [ %21, %11 ], [ %113, %91 ], [ %85, %64 ]
+      ret <4 x double> %115
+    }
+    """, "entry"), Vec{4,Float64}, Tuple{Vec{4,Float64}}, v)
+end
+
+@inline function verf(v::Vec{2,Float64})
```
Comment (chriselrod) [inline src/llvm_intrin/verf.jl:294]: mind adding a `if SSE4_1` around this definition?

File: src/llvm_intrin/verf.jl [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+      %116 = select <2 x i1> %24, <2 x double> %95, <2 x double> %115
+      br label %117
+    ; 117:                                              ; preds = %10, %68, %94
+      %118 = phi <2 x double> [ %20, %10 ], [ %116, %94 ], [ %89, %68 ]
+      ret <2 x double> %118
+    }
+    """, "entry"), Vec{2,Float64}, Tuple{Vec{2,Float64}}, v)
+end
+
+@inline function verf(v::Vec{16,Float32})
```
Comment (chriselrod) [inline src/llvm_intrin/verf.jl:432]: Similarly could you add instruction-set checks around the `Float32` definitions?

File: src/llvm_intrin/verf.jl (hunk 1 above)
Comment (chriselrod) [reply src/llvm_intrin/verf.jl:32]: This matches what the LLVM 11 language reference says: https://releases.llvm.org/11.0.0/docs/LangRef.html#function-attributes

**Comment by brenhinkeller** [comment]
I haven't added any tests yet since I'm not sure how you usually like to do that; my first thought would probably be something simple along the lines of

```
erfs = [0.1124629160182849, 0.22270258921047847, 0.3286267594591274, 0.42839235504666845, 0.5204998778130465, 0.6038560908479259, 0.6778011938374184, 0.7421009647076605, 0.7969082124228322, 0.8427007929497149, 0.8802050695740817, 0.9103139782296353, 0.9340079449406524, 0.9522851197626487, 0.9661051464753108, 0.976348383344644, 0.9837904585907745, 0.9890905016357308, 0.9927904292352575, 0.9953222650189527, 0.997020533343667, 0.9981371537020182, 0.9988568234026434, 0.999311486103355, 0.999593047982555, 0.9997639655834707, 0.9998656672600594, 0.9999249868053346, 0.9999589021219005, 0.9999779095030014, 0.9999883513426328, 0.9999939742388483]

v = verf(Vec{8, Float64}(0.1:0.1:0.8...,))
@test all([v(i) for i in 1:8] .≈ erfs[1:8])

v = verf(Vec{4, Float64}(0.1:0.1:0.4...,))
@test all([v(i) for i in 1:4] .≈ erfs[1:4])

v = verf(Vec{2, Float64}(0.1:0.1:0.2...,))
@test all([v(i) for i in 1:2] .≈ erfs[1:2])

v = verf(Vec{16, Float32}(0.1:0.1:1.6...,))
@test all([v(i) for i in 1:16] .≈ erfs[1:16])

v = verf(Vec{8, Float32}(0.1:0.1:0.8...,))
@test all([v(i) for i in 1:8] .≈ erfs[1:8])
```
but there may be a much better way to do it

**Comment by chriselrod** [comment]
You could also add a test dependency on SpecialFunctions, but a precomputed table of correct results is good and would probably result in slightly faster tests.

Also, why not the simpler?
```julia
@test [v(i) for i in 1:8] ≈ erfs[1:8]
```
These tests and the definitions should all have instruction-set tests, i.e. I think the definitions require AVX512F, AVX, and SSE4_1. These are all `Bool`s defined in VectorizationBase.

I can make sure they're defined as `false` on ARM CPUs before issuing a new release (I think they're just undefined currently). I guess you could also use `isdefined(VectorizationBase, :AVX512F) && AVX512F`, but I think it'd be nicer to just ensure they're defined.

**Comment by chriselrod** [comment]
Also, `test/testsetup.jl` includes the function `tovector`. It was written when converting to vectors was less convenient than it is now, but might still be helpful.

When working on tests, I normally `include(joinpath(pkgdir(VectorizationBase), "test", "testsetup.jl"))`, which should make it possible to and paste any test set from `test/runtests.jl`.

**Comment by brenhinkeller** [comment]
Added the tests, I'll try to work on getting them to all pass sometime tomorrow. One thing that comes to mind is that the changes to llvmcall probably mean that these new versions may not work on pre-1.6?

**Comment by chriselrod** [comment]
> Added the tests, I'll try to work on getting them to all pass sometime tomorrow. One thing that comes to mind is that the changes to llvmcall probably mean that these new versions may not work on pre-1.6?

Yes.
Two options:
1. use [llvmcall_expr](https://github.com/chriselrod/VectorizationBase.jl/blob/5a74237f5d96c6405d486bdf6bafc4e7d2d710f8/src/llvm_types.jl#L203) to create the function definitions. It takes the old-style `decl` and `instr` arguments. You'd have to use the regexes to replace `%5` and friends with other names, e.g. `%x5` for compatibility.
2. Add Julia version checks, e.g.
`if VERSION ≥ v"1.6.0-DEV.674" && SSE4_1`
and just not bother supporting it for Julia 1.5.

File: src/llvm_intrin/verf.jl (hunk 1 above)
Comment (brenhinkeller) [reply src/llvm_intrin/verf.jl]: I removed the `speculatable`s everywhere, let's see if that fixes it

File: src/llvm_intrin/verf.jl (hunk 2 above)
Comment (brenhinkeller) [reply src/llvm_intrin/verf.jl]: 👍

File: src/llvm_intrin/verf.jl (hunk 3 above)
Comment (brenhinkeller) [reply src/llvm_intrin/verf.jl]: 👍

File: src/llvm_intrin/verf.jl (hunk 4 above)
Comment (brenhinkeller) [reply src/llvm_intrin/verf.jl:29]: 👍

File: src/llvm_intrin/verf.jl (hunk 5 above)
Comment (brenhinkeller) [reply src/llvm_intrin/verf.jl:157]: 👍

File: src/llvm_intrin/verf.jl (hunk 6 above)
Comment (brenhinkeller) [reply src/llvm_intrin/verf.jl]: 👍

File: src/llvm_intrin/verf.jl (hunk 7 above)
Comment (brenhinkeller) [reply src/llvm_intrin/verf.jl]: 👍

**Comment by brenhinkeller** [comment]
Just added some quick tests for version and instruction set. I suppose anyone on pre-1.6 could still use the original SIMDPirates verf, but I'm open to the other option too.

**Comment by brenhinkeller** [comment]
Oh, looks like SSE4_1 is not defined, at least where I'm using it. Is there an obvious fix for that?

**Comment by chriselrod** [comment]
In the test suite?
```julia
julia> SSE4_1
ERROR: UndefVarError: SSE4_1 not defined

julia> VectorizationBase.SSE4_1
true
```

**Comment by brenhinkeller** [comment]
Oh that makes sense

**Comment by brenhinkeller** [comment]
Thinking about it more, it would actually make the upgrade path for my own packages a lot nicer if I could use this verf on both 1.5 and 1.6, so I'll try out llvmcall_expr later

**Comment by DilumAluthge** [comment]
CI is still failing even after https://github.com/chriselrod/VectorizationBase.jl/pull/36/commits/2a2411163681eb5cd78a63e6eb058e26c694a733

File: test/runtests.jl [hunk 8]
```diff
@@ -669,6 +669,27 @@ include("testsetup.jl")
             @test tovector(@inferred(f(vu64_1,vu64_2,vu64_3))) == f.(xu1, xu2, xu3)
         end
     end
+    @time @testset "Special functions" begin
+        if VERSION ≥ v"1.6.0-DEV.674" && SSE4_1
```
Comment (chriselrod) [inline test/runtests.jl:673]: Failing on this line.

**Comment by brenhinkeller** [comment]
Ah oops. Late night!

File: test/runtests.jl (hunk 8 above)
Comment (brenhinkeller) [reply test/runtests.jl]: 👍

**Comment by brenhinkeller** [comment]
I should probably also be checking this out and running tests locally before committing rather than trying to develop in the github browser UI :p

**Comment by brenhinkeller** [comment]
Alright, tests passing! I'll try out the `llvmcall_expr` option out either later tonight if I'm alert enough, or maybe more likely tomorrow

**Comment by chriselrod** [comment]
Want me to merge now and make another PR, or do it in this one?

**Comment by brenhinkeller** [comment]
I guess we're not breaking anything so there's no harm in merging now!

**Comment by brenhinkeller** [comment]
Hooray!

**Pull request merged**